metadata-integration/java/datahub-client/src/main/java/datahub/client/rest/StringEscapeUtils.java

```diff
@@ -0,0 +1,140 @@
+/**
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package datahub.client.rest;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.util.Locale;
+
+import org.apache.commons.lang.UnhandledException;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.linkedin.common.urn.Urn;
+import com.linkedin.data.ByteString;
+import com.linkedin.data.template.JacksonDataTemplateCodec;
+import com.linkedin.mxe.GenericAspect;
+import com.linkedin.mxe.MetadataChangeProposal;
+
+import datahub.event.MetadataChangeProposalWrapper;
+
+public class StringEscapeUtils {
+
+  private StringEscapeUtils() {
+
+  }
+
+  private final static ObjectMapper OBJECT_MAPPER = new ObjectMapper()
+      .setSerializationInclusion(JsonInclude.Include.NON_NULL);
+  private final static JacksonDataTemplateCodec DATA_TEMPLATE_CODEC = new JacksonDataTemplateCodec(
+      OBJECT_MAPPER.getFactory());
+
+  public static MetadataChangeProposal convert(MetadataChangeProposalWrapper mcpw)
+      throws IOException, URISyntaxException {
+    String serializedAspect = escapeJava(DATA_TEMPLATE_CODEC.dataTemplateToString(mcpw.getAspect()));
+    MetadataChangeProposal mcp = new MetadataChangeProposal().setEntityType(mcpw.getEntityType())
+        .setAspectName(mcpw.getAspectName()).setEntityUrn(Urn.createFromString(mcpw.getEntityUrn()))
+        .setChangeType(mcpw.getChangeType());
+
+    mcp.setAspect(new GenericAspect().setContentType("application/json")
+        .setValue(ByteString.unsafeWrap(serializedAspect.getBytes(StandardCharsets.UTF_8))));
+    return mcp;
+  }
+
+  private static void escapeJavaStyleString(Writer out, String str, boolean escapeSingleQuote,
+      boolean escapeForwardSlash) throws IOException {
+    if (out == null) {
+      throw new IllegalArgumentException("The Writer must not be null");
+    } else if (str != null) {
+      int sz = str.length();
+
+      for (int i = 0; i < sz; ++i) {
+        char ch = str.charAt(i);
+        if (ch > 4095) {
```

Comment:
This feels like fairly low-level code. Are we sure we need to manually do these substitutions? Are there any off the shelf libs to aid us in this?

Reply:
This _is_ an off the shelf library (Apache), but requires a little bit of customization to get it working the way we want.

Reply:
Note the License Header 😄